app/components/election_date_input.js

```diff
@@ -0,0 +1,209 @@
+
+import "react-calendar/dist/Calendar.css";
+import { useRef } from "react";
+
+export function ElectionDateInput(props) {
+  // defaultValue: mm/dd/yyyy string or a Date object
+  // onDone: ({valid: bool, value: Date}): null
+  const {
+    defaultValue = "",
+    onDone: propOnDone = () => {},
+  } = props;
+
+  const today = new Date();
+
+  const splitDate = (mdy) => {
+    let mm_dd_yyyy = mdy
+      .split(/-|\//)
+      .map((n) => parseInt(n, 10))
+      .filter((n) => !isNaN(n));
+    if (mm_dd_yyyy[2] < 100) {
+      mm_dd_yyyy[2] += 2000;
+    }
+    return mm_dd_yyyy;
+  };
+
+  const mdyToDate = (mdy) => {
+    if (!isMdyValid(mdy)) {
+      return new Date(NaN);
+    }
+    const [month, day, year] = splitDate(mdy);
+    return new Date(year, month - 1, day);
+  };
+
+  const isMdyValid = (mdy) => {
+    const [month, day, year] = splitDate(mdy);
+    const date = new Date(year, month - 1, day);
+    const yearLength = year?.toString().length;
+    if (
+      !(yearLength === 2 || yearLength === 4) ||
+      year != date.getFullYear() ||
+      date.getMonth() !== month - 1 ||
+      date.getDate() !== day
+    ) {
+      return false;
+    }
+    return true;
+  };
+
+  const dateToMdy = (date) => {
+    const month = date.getMonth() + 1;
+    const day = date.getDate();
+    const year = date.getFullYear();
+    return `${month}/${day}/${year}`;
+  };
+
+  const defaultDate =
+    defaultValue instanceof Date ? dateToMdy(defaultValue) : defaultValue;
+
+  const [textDate, setTextDate] = useState(defaultDate);
+  const [error, setError] = useState(null);
+  const [datePickerOpen, setDatePickerOpen] = useState(false);
+  const parentEl = useRef(null);
+
+  const classes = useStyles({
+    ...props,
+    error,
+    datePickerOpen,
+  });
+
+  useEffect(() => {
+    const onNonDatepickerClick = (e) => {
+      if (
+        !parentEl.current.contains(e.target) &&
+        !e.target.className.includes(classes.datePickerPart)
+      ) {
+        setDatePickerOpen(false);
+        blurDateInput(textDate);
+      }
+    };
+    document.addEventListener("click", onNonDatepickerClick);
+    return () => {
+      document.removeEventListener("click", onNonDatepickerClick);
+    };
+  }, [textDate]);
+
+  const onInputChange = (e) => {
+    const value = e.target.value;
+    if ((value.length === 8 || value.length === 10) && !isMdyValid(value)) {
+      setError("Please enter a valid date");
+    } else {
+      setError(null);
+    }
+    setTextDate(value);
+  };
+  const onDatePickerChange = (date) => {
+    setError(null);
+    setTextDate(dateToMdy(date));
+  };
+
+  const datePickerButtonOnClick = (textDate, datePickerOpen) => {
+    blurDateInput(textDate);
+    setDatePickerOpen(!datePickerOpen);
+  };
+  const blurDateInput = (textDate) => {
+    if (textDate !== "" && !isMdyValid(textDate)) {
+      setError("Please enter a valid date");
+    }
+    propOnDone({valid: error == null, value: mdyToDate(textDate)});
```

Comment:
When I enter a date and the blur it works fine.  But when I delete the date, and then blur, onDone gets called with valid true.  Consider that the call to setError does not change the value of error until the next render.   Other than this, it looks great.